curl/src/client.cc

```diff
@@ -0,0 +1,423 @@
+            curl_slist_free_all(list);
+        }
+    }
+
+    curl_escaped_string::curl_escaped_string(curl_handle const& handle, string const& str) :
+        scoped_resource(nullptr, cleanup)
+    {
+        _resource = curl_easy_escape(handle, str.c_str(), str.size());
+        if (!_resource) {
+            throw http_exception("curl_easy_escape failed to escape string.");
+        }
+    }
+
+    void curl_escaped_string::cleanup(char const* str)
+    {
+        if (str) {
+            curl_free(const_cast<char*>(str));
+        }
+    }
+
+    client::client()
+    {
+        if (!_handle) {
+            throw http_exception("failed to create cURL handle.");
+        }
+    }
+
+    client::client(client && other)
+    {
+        *this = std::move(other);
+    }
+
+    client &client::operator=(client && other)
+    {
+        _handle = move(other._handle);
+        return *this;
+    }
+
+    response client::get(request const& req)
+    {
+        return perform(http_method::get, req);
+    }
+
+    response client::post(request const& req)
+    {
+        return perform(http_method::post, req);
+    }
+
+    response client::put(request const& req)
+    {
+        return perform(http_method::put, req);
+    }
+
+    response client::perform(http_method method, request const& req)
+    {
+        response res;
+        context ctx(req, res);
+
+        // Reset the options
+        curl_easy_reset(_handle);
+
+        // Set common options
+        auto result = curl_easy_setopt(_handle, CURLOPT_NOPROGRESS, 1);
+        if (result != CURLE_OK) {
+            throw http_request_exception(req, curl_easy_strerror(result));
+        }
+        result = curl_easy_setopt(_handle, CURLOPT_FOLLOWLOCATION, 1);
+        if (result != CURLE_OK) {
+            throw http_request_exception(req, curl_easy_strerror(result));
+        }
+
+        // Set tracing from libcurl if enabled (we don't care if this fails)
+        if (LOG_IS_DEBUG_ENABLED()) {
+            curl_easy_setopt(_handle, CURLOPT_DEBUGFUNCTION, debug);
+            curl_easy_setopt(_handle, CURLOPT_VERBOSE, 1);
+        }
+
+        // Setup the request
+        set_method(ctx, method);
+        set_url(ctx);
+        set_headers(ctx);
+        set_cookies(ctx);
+        set_body(ctx);
+        set_timeouts(ctx);
+        set_write_callbacks(ctx);
+        set_ca_info(ctx);
+        set_client_info(ctx);
+
+        // Perform the request
+        result = curl_easy_perform(_handle);
+        if (result != CURLE_OK) {
+            throw http_request_exception(req, curl_easy_strerror(result));
+        }
+
+        LOG_DEBUG("request completed (status %1%).", res.status_code());
+
+        // Set the body of the response
+        res.body(move(ctx.response_buffer));
+        return res;
+    }
+
+    void client::set_ca_cert(std::string cert_file)
+    {
+        _ca_cert = cert_file;
+    }
+
+    void client::set_client_cert(std::string client_cert, std::string client_key)
```

Comment:
To be clear: in headers one *always* qualifies the name because it is extremely bad form to introduce using statements in headers. In source files, a `using namespace std;` is not only acceptable, but I actively encourage it as it makes the source more readable, requires less typing, and generally your names should not be conflicting with anything in the std namespace. If there are ambiguities, qualifying the name then makes sense to resolve that ambiguity; otherwise it's needless.